File: domain-searcher/domain_searcher/core/filters.py
"""
Domain name filtering engine.
@analyst: Grounding source for stop-word lists.
@developer: Pure functions, no side effects. Easy to unit-test.
"""

import re
from typing import List

# --- Stop-word categories (expandable) ---
ADULT_WORDS = [
    "porn", "xxx", "sex", "nude", "naked", "mature", "milf",
    "tube", "cam", "escort", "fetish", "hentai", "erotic",
    "onlyfans", "stripper", "hookup", "booty", "busty",
]

GAMBLING_WORDS = [
    "casino", "bet", "poker", "slots", "jackpot", "gamble",
    "roulette", "blackjack", "lottery", "bingo", "wager",
    "sportbet", "betting", "1xbet", "stake",
]

DRUG_WORDS = [
    "weed", "cannabis", "marijuana", "drug", "pill", "pharma",
    "opioid", "cocaine", "meth", "heroin", "kratom", "cbd",
    "thc", "vape", "smoke", "tobacco",
]

PROFANITY_EN = [
    "fuck", "shit", "ass", "damn", "bitch", "crap",
    "dick", "cock", "cunt", "whore", "slut",
]

PROFANITY_RU = [
    "блят", "хуй", "пизд", "ебат", "сука", "мудак",
    "жоп", "дерьм", "шлюх",
]

SPAM_PATTERNS = [
    "buy-", "cheap-", "free-", "best-", "top-",
    "click", "deal", "discount", "promo",
]

# Combined master list
ALL_STOP_WORDS: List[str] = (
    ADULT_WORDS + GAMBLING_WORDS + DRUG_WORDS
    + PROFANITY_EN + PROFANITY_RU + SPAM_PATTERNS
)


def is_clean_domain(domain_name: str, extra_stop_words: List[str] | None = None) -> bool:
    """
    Check if a domain name is free of trigger/stop words.

    Args:
        domain_name: The domain to check (e.g., "example.com").
        extra_stop_words: Optional additional words to filter.

    Returns:
        True if the domain is clean, False if it contains a trigger word.
    """
    name_lower = domain_name.lower().replace(".com", "").replace(".", "")
    stop_list = ALL_STOP_WORDS + (extra_stop_words or [])

    for word in stop_list:
        if word in name_lower:
            return False

    return True


def is_valid_domain_format(domain_name: str) -> bool:
    """Basic validation that the string looks like a domain."""
    pattern = r"^[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.com$"
    return bool(re.match(pattern, domain_name))


def filter_candidates(domains: List[str], extra_stop_words: List[str] | None = None) -> List[str]:
    """
    Filter a list of domain names, removing those with stop words
    and invalid formats.

    Returns:
        List of clean, valid domain names.
    """
    return [
        d for d in domains
        if is_valid_domain_format(d) and is_clean_domain(d, extra_stop_words)
    ]

File: domain-searcher/extension_v2.5/filters.js
/**
 * Domain name filtering engine.
 * Direct port from Python filters.py — stop-word lists + validation.
 */

const ADULT_WORDS = [
    "porn", "xxx", "sex", "nude", "naked", "mature", "milf",
    "tube", "cam", "escort", "fetish", "hentai", "erotic",
    "onlyfans", "stripper", "hookup", "booty", "busty",
];

const GAMBLING_WORDS = [
    "casino", "bet", "poker", "slots", "jackpot", "gamble",
    "roulette", "blackjack", "lottery", "bingo", "wager",
    "sportbet", "betting", "1xbet", "stake",
];

const DRUG_WORDS = [
    "weed", "cannabis", "marijuana", "drug", "pill", "pharma",
    "opioid", "cocaine", "meth", "heroin", "kratom", "cbd",
    "thc", "vape", "smoke", "tobacco",
];

const PROFANITY_EN = [
    "fuck", "shit", "ass", "damn", "bitch", "crap",
    "dick", "cock", "cunt", "whore", "slut",
];

const PROFANITY_RU = [
    "\u0431\u043b\u044f\u0442", "\u0445\u0443\u0439", "\u043f\u0438\u0437\u0434", "\u0435\u0431\u0430\u0442", "\u0441\u0443\u043a\u0430", "\u043c\u0443\u0434\u0430\u043a",
    "\u0436\u043e\u043f", "\u0434\u0435\u0440\u044c\u043c", "\u0448\u043b\u044e\u0445",
];

const SPAM_PATTERNS = [
    "buy-", "cheap-", "free-", "best-", "top-",
    "click", "deal", "discount", "promo",
];

const ALL_STOP_WORDS = [
    ...ADULT_WORDS, ...GAMBLING_WORDS, ...DRUG_WORDS,
    ...PROFANITY_EN, ...PROFANITY_RU, ...SPAM_PATTERNS,
];

/**
 * Check if a domain name is free of trigger/stop words.
 * @param {string} domainName - e.g. "example.com"
 * @param {string[]} [extraStopWords] - optional additional words
 * @returns {boolean} true if the domain is clean
 */
function isCleanDomain(domainName, extraStopWords = []) {
    const nameLower = domainName.toLowerCase().replace(".com", "").replace(/\./g, "");
    const stopList = [...ALL_STOP_WORDS, ...extraStopWords];

    for (const word of stopList) {
        if (nameLower.includes(word)) {
            return false;
        }
    }
    return true;
}

/**
 * Basic validation that the string looks like a .com domain.
 * @param {string} domainName
 * @returns {boolean}
 */
function isValidDomainFormat(domainName) {
    return /^[a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.com$/.test(domainName);
}
